src/interface/swarm_device_context.hpp

```diff
@@ -0,0 +1,112 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+#define INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+
+namespace parthenon {
+
+class SwarmDeviceContext {
+ public:
+  KOKKOS_FUNCTION
+  bool IsActive(int n) const { return mask_(n); }
+
+  KOKKOS_FUNCTION
+  bool IsOnCurrentMeshBlock(int n) const { return blockIndex_(n) == this_block_; }
+
+  KOKKOS_FUNCTION
+  void MarkParticleForRemoval(int n) const { marked_for_removal_(n) = true; }
+
+  KOKKOS_FUNCTION
+  bool IsMarkedForRemoval(const int n) const { return marked_for_removal_(n); }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetNeighborBlockIndex(const int &n, const double &x, const double &y,
+                            const double &z, bool &is_on_current_mesh_block) const {
+    int i = static_cast<int>(std::floor((x - x_min_) / ((x_max_ - x_min_) / 2.))) + 1;
+    int j = static_cast<int>(std::floor((y - y_min_) / ((y_max_ - y_min_) / 2.))) + 1;
+    int k = static_cast<int>(std::floor((z - z_min_) / ((z_max_ - z_min_) / 2.))) + 1;
+
+    // Something went wrong
+    if (i < 0 || i > 3 || ((j < 0 || j > 3) && ndim_ > 1) ||
+        ((k < 0 || k > 3) && ndim_ > 2)) {
+      PARTHENON_FAIL("Particle neighbor indices out of bounds");
+    }
+
+    // Ignore k,j indices as necessary based on problem dimension
+    if (ndim_ == 1) {
+      blockIndex_(n) = neighborIndices_(0, 0, i);
+    } else if (ndim_ == 2) {
+      blockIndex_(n) = neighborIndices_(0, j, i);
+    } else {
+      blockIndex_(n) = neighborIndices_(k, j, i);
+    }
+
+    is_on_current_mesh_block = (blockIndex_(n) == this_block_);
+
+    return blockIndex_(n);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetMyRank() const { return my_rank_; }
+
+  KOKKOS_INLINE_FUNCTION
+  void Xtoijk(const Real &x, const Real &y, const Real &z, int &i, int &j, int &k) const {
+    if (ndim_ > 0) {
```

Comment:
Is there a case where `ndim_` is not `>0`?

Reply:
No -- the relevant line in `mesh/mesh.cpp`: `ndim((mesh_size.nx3 > 1) ? 3 : ((mesh_size.nx2 > 1) ? 2 : 1))`. I'll remove this conditional.